File: backend_test.py
#!/usr/bin/env python3
"""
Backend Test for PHP/SQLite Tile Inventory System
Tests database schema, stock calculations, and API endpoints
"""

import requests
import sqlite3
import sys
import os
from datetime import datetime

class TileInventoryTester:
    def __init__(self, base_url="https://inventory-tracker-159.preview.emergentagent.com"):
        self.base_url = base_url
        self.db_path = "/app/data/app.sqlite"
        self.tests_run = 0
        self.tests_passed = 0
        self.session = requests.Session()
        
    def log_test(self, name, success, message=""):
        """Log test result"""
        self.tests_run += 1
        if success:
            self.tests_passed += 1
            print(f"✅ {name}: PASSED {message}")
        else:
            print(f"❌ {name}: FAILED {message}")
        return success

    def test_database_connection(self):
        """Test SQLite database connection and basic structure"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Check if database file exists and is accessible
            cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
            tables = [row[0] for row in cursor.fetchall()]
            
            required_tables = ['tiles', 'misc_items', 'quotations', 'quotation_items', 
                             'quotation_misc_items', 'purchase_entries_tiles', 'purchase_entries_misc']
            
            missing_tables = [table for table in required_tables if table not in tables]
            
            if missing_tables:
                return self.log_test("Database Connection", False, f"Missing tables: {missing_tables}")
            
            conn.close()
            return self.log_test("Database Connection", True, f"Found {len(tables)} tables")
            
        except Exception as e:
            return self.log_test("Database Connection", False, f"Error: {str(e)}")

    def test_active_column_migration(self):
        """Test if active column was added to tiles and misc_items tables"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Check tiles table for active column
            cursor.execute("PRAGMA table_info(tiles)")
            tiles_columns = [col[1] for col in cursor.fetchall()]
            tiles_has_active = 'active' in tiles_columns
            
            # Check misc_items table for active column
            cursor.execute("PRAGMA table_info(misc_items)")
            misc_columns = [col[1] for col in cursor.fetchall()]
            misc_has_active = 'active' in misc_columns
            
            conn.close()
            
            if tiles_has_active and misc_has_active:
                return self.log_test("Active Column Migration", True, "Both tables have active column")
            else:
                missing = []
                if not tiles_has_active:
                    missing.append("tiles")
                if not misc_has_active:
                    missing.append("misc_items")
                return self.log_test("Active Column Migration", False, f"Missing active column in: {missing}")
                
        except Exception as e:
            return self.log_test("Active Column Migration", False, f"Error: {str(e)}")

    def test_database_views(self):
        """Test if current_tiles_stock and current_misc_stock views exist"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Check for views
            cursor.execute("SELECT name FROM sqlite_master WHERE type='view'")
            views = [row[0] for row in cursor.fetchall()]
            
            required_views = ['current_tiles_stock', 'current_misc_stock']
            missing_views = [view for view in required_views if view not in views]
            
            if missing_views:
                conn.close()
                return self.log_test("Database Views", False, f"Missing views: {missing_views}")
            
            # Test if views return data
            cursor.execute("SELECT COUNT(*) FROM current_tiles_stock")
            tiles_count = cursor.fetchone()[0]
            
            cursor.execute("SELECT COUNT(*) FROM current_misc_stock")
            misc_count = cursor.fetchone()[0]
            
            conn.close()
            return self.log_test("Database Views", True, f"Views exist - Tiles: {tiles_count}, Misc: {misc_count}")
            
        except Exception as e:
            return self.log_test("Database Views", False, f"Error: {str(e)}")

    def test_web_server_response(self):
        """Test if web server is responding"""
        try:
            response = self.session.get(f"{self.base_url}/login.php", timeout=10)
            if response.status_code == 200:
                return self.log_test("Web Server Response", True, f"Status: {response.status_code}")
            else:
                return self.log_test("Web Server Response", False, f"Status: {response.status_code}")
        except Exception as e:
            return self.log_test("Web Server Response", False, f"Error: {str(e)}")

    def test_quotation_enhanced_page(self):
        """Test if quotation_enhanced.php loads without errors"""
        try:
            # First try to access login page to get session
            login_response = self.session.get(f"{self.base_url}/login.php")
            
            # Try to access quotation page (might redirect to login)
            response = self.session.get(f"{self.base_url}/quotation_enhanced.php", timeout=10)
            
            if response.status_code == 200:
                # Check if page contains expected content
                content = response.text.lower()
                if "quotation" in content and "create" in content:
                    return self.log_test("Quotation Enhanced Page", True, "Page loads with expected content")
                else:
                    return self.log_test("Quotation Enhanced Page", False, "Page loads but missing expected content")
            else:
                return self.log_test("Quotation Enhanced Page", False, f"Status: {response.status_code}")
                
        except Exception as e:
            return self.log_test("Quotation Enhanced Page", False, f"Error: {str(e)}")

    def test_quotation_view_page(self):
        """Test if quotation_view.php exists and loads"""
        try:
            # Test with a dummy ID - should either show quotation or redirect/error gracefully
            response = self.session.get(f"{self.base_url}/quotation_view.php?id=1", timeout=10)
            
            if response.status_code == 200:
                return self.log_test("Quotation View Page", True, "Page exists and loads")
            elif response.status_code == 302:
                return self.log_test("Quotation View Page", True, "Page exists (redirected)")
            else:
                return self.log_test("Quotation View Page", False, f"Status: {response.status_code}")
                
        except Exception as e:
            return self.log_test("Quotation View Page", False, f"Error: {str(e)}")

    def test_other_purchase_page(self):
        """Test if other_purchase.php loads with enhanced functionality"""
        try:
            response = self.session.get(f"{self.base_url}/other_purchase.php", timeout=10)
            
            if response.status_code == 200:
                content = response.text.lower()
                # Check for hide/show functionality indicators
                if "active" in content and "hide" in content:
                    return self.log_test("Other Purchase Page", True, "Page loads with hide/show functionality")
                else:
                    return self.log_test("Other Purchase Page", True, "Page loads but hide/show functionality unclear")
            else:
                return self.log_test("Other Purchase Page", False, f"Status: {response.status_code}")
                
        except Exception as e:
            return self.log_test("Other Purchase Page", False, f"Error: {str(e)}")

    def test_inventory_report_page(self):
        """Test if report_inventory_enhanced.php loads"""
        try:
            response = self.session.get(f"{self.base_url}/report_inventory_enhanced.php", timeout=10)
            
            if response.status_code == 200:
                return self.log_test("Inventory Report Page", True, "Page loads successfully")
            elif response.status_code == 302:
                return self.log_test("Inventory Report Page", True, "Page exists (redirected - likely auth)")
            else:
                return self.log_test("Inventory Report Page", False, f"Status: {response.status_code}")
                
        except Exception as e:
            return self.log_test("Inventory Report Page", False, f"Error: {str(e)}")

    def test_daily_pl_report_page(self):
        """Test if report_daily_pl.php loads"""
        try:
            response = self.session.get(f"{self.base_url}/report_daily_pl.php", timeout=10)
            
            if response.status_code == 200:
                return self.log_test("Daily P&L Report Page", True, "Page loads successfully")
            elif response.status_code == 302:
                return self.log_test("Daily P&L Report Page", True, "Page exists (redirected - likely auth)")
            else:
                return self.log_test("Daily P&L Report Page", False, f"Status: {response.status_code}")
                
        except Exception as e:
            return self.log_test("Daily P&L Report Page", False, f"Error: {str(e)}")

    def test_stock_calculation_consistency(self):
        """Test if stock calculations are consistent across views"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Test tiles stock calculation
            cursor.execute("""
                SELECT t.id, t.name, 
                       COALESCE(cts.total_stock_boxes, 0) as view_stock,
                       COALESCE(SUM(pe.usable_boxes), 0) as direct_stock
                FROM tiles t
                LEFT JOIN current_tiles_stock cts ON t.id = cts.id
                LEFT JOIN purchase_entries_tiles pe ON t.id = pe.tile_id
                WHERE t.active = 1
                GROUP BY t.id, t.name, cts.total_stock_boxes
                LIMIT 5
            """)
            
            tiles_results = cursor.fetchall()
            tiles_consistent = True
            
            for tile in tiles_results:
                if abs(tile[2] - tile[3]) > 0.01:  # Allow small floating point differences
                    tiles_consistent = False
                    break
            
            # Test misc items stock calculation
            cursor.execute("""
                SELECT m.id, m.name,
                       COALESCE(cms.total_stock_quantity, 0) as view_stock,
                       COALESCE(SUM(pe.usable_quantity), 0) as direct_stock
                FROM misc_items m
                LEFT JOIN current_misc_stock cms ON m.id = cms.id
                LEFT JOIN purchase_entries_misc pe ON m.id = pe.misc_item_id
                WHERE m.active = 1
                GROUP BY m.id, m.name, cms.total_stock_quantity
                LIMIT 5
            """)
            
            misc_results = cursor.fetchall()
            misc_consistent = True
            
            for item in misc_results:
                if abs(item[2] - item[3]) > 0.01:  # Allow small floating point differences
                    misc_consistent = False
                    break
            
            conn.close()
            
            if tiles_consistent and misc_consistent:
                return self.log_test("Stock Calculation Consistency", True, 
                                   f"Tested {len(tiles_results)} tiles, {len(misc_results)} misc items")
            else:
                return self.log_test("Stock Calculation Consistency", False, 
                                   f"Inconsistencies found in stock calculations")
                
        except Exception as e:
            return self.log_test("Stock Calculation Consistency", False, f"Error: {str(e)}")

    def run_all_tests(self):
        """Run all backend tests"""
        print("🔍 Starting Backend Tests for PHP/SQLite Tile Inventory System")
        print("=" * 60)
        
        # Database tests
        self.test_database_connection()
        self.test_active_column_migration()
        self.test_database_views()
        self.test_stock_calculation_consistency()
        
        # Web server tests
        self.test_web_server_response()
        self.test_quotation_enhanced_page()
        self.test_quotation_view_page()
        self.test_other_purchase_page()
        self.test_inventory_report_page()
        self.test_daily_pl_report_page()
        
        print("=" * 60)
        print(f"📊 Backend Tests Summary: {self.tests_passed}/{self.tests_run} passed")
        
        if self.tests_passed == self.tests_run:
            print("🎉 All backend tests passed!")
            return 0
        else:
            print(f"⚠️  {self.tests_run - self.tests_passed} tests failed")
            return 1

def main():
    tester = TileInventoryTester()
    return tester.run_all_tests()

if __name__ == "__main__":
    sys.exit(main())

File: backend_test.php
<?php
// backend_test.php - Comprehensive Backend Testing for Inventory System
require_once __DIR__ . '/includes/Database.php';
require_once __DIR__ . '/includes/inventory_updates.php';
require_once __DIR__ . '/includes/inventory_calculations.php';

class InventorySystemTester {
    private $pdo;
    private $tests_run = 0;
    private $tests_passed = 0;
    private $test_results = [];
    
    public function __construct() {
        try {
            $this->pdo = Database::pdo();
            echo "✅ Database connection established\n";
        } catch (Exception $e) {
            echo "❌ Database connection failed: " . $e->getMessage() . "\n";
            exit(1);
        }
    }
    
    public function runAllTests() {
        echo "\n🔍 Starting Inventory System Backend Tests...\n";
        echo "=" . str_repeat("=", 50) . "\n\n";
        
        // Test database structure
        $this->testDatabaseStructure();
        
        // Test inventory calculations
        $this->testInventoryCalculations();
        
        // Test inventory updates (sales/returns)
        $this->testInventoryUpdates();
        
        // Test inventory views and summaries
        $this->testInventoryViews();
        
        // Test data integrity
        $this->testDataIntegrity();
        
        $this->printSummary();
        return $this->tests_passed === $this->tests_run;
    }
    
    private function runTest($name, $callback) {
        $this->tests_run++;
        echo "🔍 Testing: $name... ";
        
        try {
            $result = $callback();
            if ($result) {
                $this->tests_passed++;
                echo "✅ PASSED\n";
                $this->test_results[] = ['name' => $name, 'status' => 'PASSED', 'message' => ''];
            } else {
                echo "❌ FAILED\n";
                $this->test_results[] = ['name' => $name, 'status' => 'FAILED', 'message' => 'Test returned false'];
            }
        } catch (Exception $e) {
            echo "❌ ERROR: " . $e->getMessage() . "\n";
            $this->test_results[] = ['name' => $name, 'status' => 'ERROR', 'message' => $e->getMessage()];
        }
    }
    
    private function testDatabaseStructure() {
        echo "\n📋 Testing Database Structure...\n";
        
        // Test required tables exist
        $this->runTest("Required tables exist", function() {
            $required_tables = [
                'tiles', 'tile_sizes', 'inventory_items', 'invoices', 'invoice_items', 
                'individual_returns', 'purchase_entries_tiles'
            ];
            
            foreach ($required_tables as $table) {
                $stmt = $this->pdo->prepare("SELECT name FROM sqlite_master WHERE type='table' AND name = ?");
                $stmt->execute([$table]);
                if (!$stmt->fetchColumn()) {
                    throw new Exception("Required table '$table' not found");
                }
            }
            return true;
        });
        
        // Test inventory views can be created
        $this->runTest("Inventory views creation", function() {
            InventoryUpdates::createInventoryViews($this->pdo);
            
            // Check if view exists
            $stmt = $this->pdo->prepare("SELECT name FROM sqlite_master WHERE type='view' AND name = 'current_inventory_summary'");
            $stmt->execute();
            return (bool)$stmt->fetchColumn();
        });
    }
    
    private function testInventoryCalculations() {
        echo "\n🧮 Testing Inventory Calculations...\n";
        
        // Test basic cost calculations
        $this->runTest("Basic cost calculations", function() {
            $test_data = [
                'boxes_in' => 100,
                'damage_boxes' => 5,
                'per_box_value' => 50,
                'transport_pct' => 10,
                'transport_per_box' => 2
            ];
            
            $result = InventoryCalculations::calculateItemCosts($test_data, 1.5);
            
            // Verify calculations
            if ($result['net_boxes'] !== 95.0) {
                throw new Exception("Net boxes calculation incorrect: expected 95, got " . $result['net_boxes']);
            }
            
            if ($result['damage_percentage'] !== 5.0) {
                throw new Exception("Damage percentage calculation incorrect: expected 5%, got " . $result['damage_percentage'] . "%");
            }
            
            // Expected: base 50 + transport (50*0.1 + 2) = 50 + 5 + 2 = 57
            if ($result['final_cost_per_box'] !== 57.0) {
                throw new Exception("Final cost per box incorrect: expected 57, got " . $result['final_cost_per_box']);
            }
            
            return true;
        });
        
        // Test validation
        $this->runTest("Data validation", function() {
            $invalid_data = [
                'tile_id' => 0,
                'boxes_in' => -10,
                'damage_boxes' => 150,
                'per_box_value' => -5
            ];
            
            $validation = InventoryCalculations::validateInventoryData($invalid_data);
            
            if ($validation['valid']) {
                throw new Exception("Validation should have failed for invalid data");
            }
            
            if (count($validation['errors']) < 3) {
                throw new Exception("Expected multiple validation errors, got " . count($validation['errors']));
            }
            
            return true;
        });
    }
    
    private function testInventoryUpdates() {
        echo "\n🔄 Testing Inventory Updates (Sales/Returns)...\n";
        
        // Test sales processing
        $this->runTest("Sales inventory deduction", function() {
            // Create test data
            $test_tile_id = $this->createTestTile();
            $this->addTestInventory($test_tile_id, 100); // Add 100 boxes
            
            // Get initial inventory
            $initial_levels = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            $initial_stock = 0;
            foreach ($initial_levels as $item) {
                if ($item['tile_id'] == $test_tile_id) {
                    $initial_stock = (float)$item['available_boxes'];
                    break;
                }
            }
            
            // Process a sale
            $sale_items = [
                ['tile_id' => $test_tile_id, 'boxes_decimal' => 25]
            ];
            InventoryUpdates::processSale($this->pdo, 999, $sale_items);
            
            // Check inventory after sale
            $updated_levels = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            $updated_stock = 0;
            foreach ($updated_levels as $item) {
                if ($item['tile_id'] == $test_tile_id) {
                    $updated_stock = (float)$item['available_boxes'];
                    break;
                }
            }
            
            // Should be reduced by 25
            $expected_stock = $initial_stock - 25;
            if (abs($updated_stock - $expected_stock) > 0.01) {
                throw new Exception("Stock not properly deducted. Expected: $expected_stock, Got: $updated_stock");
            }
            
            return true;
        });
        
        // Test returns processing
        $this->runTest("Returns inventory restoration", function() {
            $test_tile_id = $this->createTestTile();
            $this->addTestInventory($test_tile_id, 50);
            
            // Get initial stock
            $initial_levels = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            $initial_stock = 0;
            foreach ($initial_levels as $item) {
                if ($item['tile_id'] == $test_tile_id) {
                    $initial_stock = (float)$item['available_boxes'];
                    break;
                }
            }
            
            // Process a return
            $return_data = [
                'tile_id' => $test_tile_id,
                'quantity' => 10
            ];
            InventoryUpdates::processReturn($this->pdo, 888, $return_data);
            
            // Check inventory after return
            $updated_levels = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            $updated_stock = 0;
            foreach ($updated_levels as $item) {
                if ($item['tile_id'] == $test_tile_id) {
                    $updated_stock = (float)$item['available_boxes'];
                    break;
                }
            }
            
            // Should be increased by 10
            $expected_stock = $initial_stock + 10;
            if (abs($updated_stock - $expected_stock) > 0.01) {
                throw new Exception("Stock not properly restored. Expected: $expected_stock, Got: $updated_stock");
            }
            
            return true;
        });
    }
    
    private function testInventoryViews() {
        echo "\n📊 Testing Inventory Views and Summaries...\n";
        
        // Test inventory summary data
        $this->runTest("Inventory summary data structure", function() {
            $inventory_data = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            
            if (!is_array($inventory_data)) {
                throw new Exception("Inventory data should be an array");
            }
            
            // Check required fields in each item
            if (!empty($inventory_data)) {
                $required_fields = [
                    'tile_name', 'size_label', 'total_received', 'total_sold', 
                    'total_returned', 'available_boxes', 'current_stock', 
                    'total_cost_value', 'avg_cost_per_box'
                ];
                
                $first_item = $inventory_data[0];
                foreach ($required_fields as $field) {
                    if (!array_key_exists($field, $first_item)) {
                        throw new Exception("Required field '$field' missing from inventory data");
                    }
                }
            }
            
            return true;
        });
        
        // Test reconciliation
        $this->runTest("Inventory reconciliation", function() {
            $result = InventoryUpdates::reconcileInventory($this->pdo);
            return $result === true;
        });
    }
    
    private function testDataIntegrity() {
        echo "\n🔒 Testing Data Integrity...\n";
        
        // Test negative stock prevention
        $this->runTest("Negative stock prevention", function() {
            $test_tile_id = $this->createTestTile();
            $this->addTestInventory($test_tile_id, 10); // Only 10 boxes
            
            // Try to sell more than available
            $sale_items = [
                ['tile_id' => $test_tile_id, 'boxes_decimal' => 50] // Try to sell 50
            ];
            
            // This should still process but result in negative available stock
            InventoryUpdates::processSale($this->pdo, 777, $sale_items);
            
            $levels = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            $stock = 0;
            foreach ($levels as $item) {
                if ($item['tile_id'] == $test_tile_id) {
                    $stock = (float)$item['available_boxes'];
                    break;
                }
            }
            
            // Should be negative (10 - 50 = -40)
            if ($stock >= 0) {
                throw new Exception("System should allow negative stock for tracking purposes, got: $stock");
            }
            
            return true;
        });
        
        // Test cost calculations consistency
        $this->runTest("Cost calculations consistency", function() {
            $inventory_data = InventoryUpdates::getCurrentInventoryLevels($this->pdo);
            
            foreach ($inventory_data as $item) {
                $available_boxes = (float)$item['available_boxes'];
                $avg_cost = (float)$item['avg_cost_per_box'];
                $total_cost = (float)$item['total_cost_value'];
                
                if ($available_boxes > 0) {
                    $calculated_total = $available_boxes * $avg_cost;
                    if (abs($calculated_total - $total_cost) > 0.01) {
                        throw new Exception("Cost calculation inconsistency for tile {$item['tile_name']}: expected $calculated_total, got $total_cost");
                    }
                }
            }
            
            return true;
        });
    }
    
    private function createTestTile() {
        // Create a test tile size if not exists
        $this->pdo->exec("INSERT OR IGNORE INTO tile_sizes (id, label, sqft_per_box) VALUES (999, 'Test Size', 1.5)");
        
        // Create a test tile with unique name
        $unique_name = 'Test Tile ' . time() . '_' . rand(1000, 9999);
        $stmt = $this->pdo->prepare("INSERT INTO tiles (name, size_id) VALUES (?, ?)");
        $stmt->execute([$unique_name, 999]);
        
        return (int)$this->pdo->lastInsertId();
    }
    
    private function addTestInventory($tile_id, $boxes) {
        $stmt = $this->pdo->prepare("
            INSERT INTO inventory_items (tile_id, purchase_dt, boxes_in, per_box_value, vendor)
            VALUES (?, ?, ?, ?, ?)
        ");
        $stmt->execute([$tile_id, date('Y-m-d'), $boxes, 50, 'Test Vendor']);
    }
    
    private function printSummary() {
        echo "\n" . str_repeat("=", 60) . "\n";
        echo "📊 TEST SUMMARY\n";
        echo str_repeat("=", 60) . "\n";
        echo "Total Tests: {$this->tests_run}\n";
        echo "Passed: {$this->tests_passed}\n";
        echo "Failed: " . ($this->tests_run - $this->tests_passed) . "\n";
        echo "Success Rate: " . round(($this->tests_passed / $this->tests_run) * 100, 1) . "%\n";
        
        if ($this->tests_passed === $this->tests_run) {
            echo "\n🎉 ALL TESTS PASSED! Inventory system is working correctly.\n";
        } else {
            echo "\n⚠️  SOME TESTS FAILED. Check the details above.\n";
            
            echo "\nFailed Tests:\n";
            foreach ($this->test_results as $result) {
                if ($result['status'] !== 'PASSED') {
                    echo "- {$result['name']}: {$result['status']}";
                    if ($result['message']) {
                        echo " ({$result['message']})";
                    }
                    echo "\n";
                }
            }
        }
        echo str_repeat("=", 60) . "\n";
    }
}

// Run the tests
$tester = new InventorySystemTester();
$success = $tester->runAllTests();

exit($success ? 0 : 1);
?>